### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by nest1ing**
This is an initial version of project provides the base functionality.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -5,11 +5,61 @@
...
+    source.get_event().exit(EXIT_SUCCESS);
+}
 
 int main([[maybe_unused]] int argc, [[maybe_unused]]char* argv[])
 {
-    printf("obmc-yadro-fssync ver %s\n", PROJECT_VERSION);
+    fmt::print("obmc-yadro-fssync ver {}\n", PROJECT_VERSION);
+
+    auto event = sdeventplus::Event::get_default();
+    int rc = EXIT_SUCCESS;
```
Comment (AlexanderAmelkin) [inline src/main.cpp]: Generally it is better to default to failure rather than success, but here I don't insist.

File: src/watch.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+{
+  public:
+    Watch() = delete;
+    Watch(const Watch&) = delete;
+    Watch& operator=(const Watch&) = delete;
+    Watch(Watch&&) = delete;
+    Watch& operator=(Watch&&) = delete;
+
+    /**
+     * @brief dtor - remove inotify watch adn close fd's
```
Comment (AlexanderAmelkin) [inline src/watch.hpp]: `adn` -> `and` ?

File: src/watch.cpp [hunk 3]
```diff
@@ -146,6 +147,15 @@ void Watch::rescanRoot(sdeventplus::source::EventBase&)
     addWatch(root);
 }
 
+void Watch::checkWds(sdeventplus::source::EventBase& source)
+{
+    if (wds.empty())
+    {
+        fmt::print(stderr, "ERROR: No directories to watch exist.\n");
+        source.get_event().exit(ENXIO);
```
Comment (AlexanderAmelkin) [inline src/watch.cpp]: Would be probably better to return `ENOTDIR` or `ENOENT` ?

File: src/whitelist.hpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace fssync
+{
+
+namespace details
+{
+struct Pred
+{
+    /**
+     * @brief Compare the begin of strings.
```
Comment (AlexanderAmelkin) [inline src/whitelist.hpp]: `Compare the beginnings of strings`

File: src/whitelist.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#include <set>
+
+namespace fs = std::filesystem;
+
+namespace fssync
+{
+
+namespace details
+{
+struct Pred
```
Comment (AlexanderAmelkin) [inline src/whitelist.hpp]: What does `Pred` stand for? Did you mean `Prefix` ?

File: src/whitelist.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+     *
+     * For example:
+     * The comparison of the strings like `etc/systemd` and
+     * `etc/systemd/network` should return true.
+     */
+    bool operator()(const std::string& lhs, const std::string& rhs) const
+    {
+        const auto llen = lhs.length();
+        const auto rlen = rhs.length();
+        return lhs.compare(0, llen, rhs, 0, std::min(llen, rlen)) < 0;
```
Comment (AlexanderAmelkin) [inline src/whitelist.hpp:30]: I guess with C++20 you could use just `rhs.starts_with(lhs)` ?

File: src/main.cpp (hunk 1 above)
Comment (nest1ing) [reply src/main.cpp]: Fixed

File: src/watch.hpp (hunk 2 above)
Comment (nest1ing) [reply src/watch.hpp]: Fixed

File: src/watch.cpp (hunk 3 above)
Comment (nest1ing) [reply src/watch.cpp]: Fixed

File: src/whitelist.hpp (hunk 4 above)
Comment (nest1ing) [reply src/whitelist.hpp]: Fixed

File: src/whitelist.hpp (hunk 5 above)
Comment (nest1ing) [reply src/whitelist.hpp]: I meant `Predicate`. Renamed to `PathsComparer`.

File: src/whitelist.hpp (hunk 6 above)
Comment (nest1ing) [reply src/whitelist.hpp:30]: Until we catch up the upstream in our internal repo we have to use C++17.  And this code is designed exactly for C++17.

File: src/main.cpp [hunk 7]
```diff
@@ -5,11 +5,171 @@
...
+        app);
+    fmt::print(R"(Required arguments:
+  source-dir            Path to the source directory.
+  dest-dir              Path to the destination directory.
+
+Optional arguments:
+  -h, --help            show this help message and exit.
+  -d, --delay SECONDS   define delay before sync process starting
+  -w, --witelist FILE   path to the file with a list of interesting files.
+                        File should contain a pathes relative to source-dri.
```
Comment (AlexanderAmelkin) [inline src/main.cpp]: `a pathes` -> `paths`.

File: src/main.cpp [hunk 8]
```diff
@@ -5,11 +5,171 @@
...
+        "\nUsage: {} [-h] [-d SECONDS] [-w FILE] <source-dir> <dest-dir>\n",
+        app);
+    fmt::print(R"(Required arguments:
+  source-dir            Path to the source directory.
+  dest-dir              Path to the destination directory.
+
+Optional arguments:
+  -h, --help            show this help message and exit.
+  -d, --delay SECONDS   define delay before sync process starting
+  -w, --witelist FILE   path to the file with a list of interesting files.
```
Comment (AlexanderAmelkin) [inline src/main.cpp]: `a list of files to track` maybe?

File: src/main.cpp [hunk 9]
```diff
@@ -5,11 +5,171 @@
...
+                        File should contain a pathes relative to source-dri.
+                        If not specified all files from source directory
```
Comment (AlexanderAmelkin) [inline src/main.cpp]: `, all`, `the source`.

File: src/watch.hpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                        Callback callback);
+
+  protected:
+    /**
+     * @brief ctor - hook inotify watch with sd-event
+     *
+     * @param event - sd-event object
+     * @param fd    - inotify fd
+     * @param root  - root directory watched to
+     * @param callback - The callback function for processing files
```
Comment (AlexanderAmelkin) [inline src/watch.hpp:48]: Strange alignment. Please either align them all or don't align them all, not this strange mix. :)

File: src/whitelist.hpp [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace fssync
+{
+
+namespace details
+{
+struct PathsComparer
+{
+    /**
+     * @brief Compare the begining of strings.
```
Comment (AlexanderAmelkin) [inline src/whitelist.hpp]: still, `beginNingS` :)

File: src/whitelist.hpp [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+     *
+     * For example:
+     * The comparison of the strings like `etc/systemd` and
```
Comment (AlexanderAmelkin) [inline src/whitelist.hpp]: No first `The` please. The second `the` is optional as `strings` is plural.

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 7 above)
Comment (nest1ing) [reply src/main.cpp]: fixed

File: src/main.cpp (hunk 8 above)
Comment (nest1ing) [reply src/main.cpp]: fixed

File: src/main.cpp (hunk 9 above)
Comment (nest1ing) [reply src/main.cpp]: fixed

File: src/watch.hpp (hunk 10 above)
Comment (nest1ing) [reply src/watch.hpp:48]: fixed

File: src/whitelist.hpp (hunk 11 above)
Comment (nest1ing) [reply src/whitelist.hpp]: refixed :)

File: src/whitelist.hpp (hunk 12 above)
Comment (nest1ing) [reply src/whitelist.hpp]: fixed

**Review by AlexanderAmelkin — APPROVED** [review]

**Pull request merged**